## CS431/631 Data Intensive Distributed Computing
### Winter 2020 - Assignment 5
---

**Please edit this (text) cell to provide your name and UW student ID number!**
* **Name:** _Kamran Karim_
* **ID:** _20594936_

In [3]:
import findspark, random
findspark.init("/u/cs451/packages/spark")

from pyspark import SparkContext, SparkConf
sc = SparkContext(appName="YourTest", master="local[2]", conf=SparkConf().set('spark.ui.port', random.randrange(4000,5000)))

---
#### Overview
For this assignment, you will be using Python and Spark to perform spam detection.   You will need to perform two tasks.   The first is to build spam prediction models, using training data sets and stochastic gradient descent (SGD).   The second is to use these models to predict whether the documents in a test data set are spam.
The stochastic gradient descent technique that you will be using is based on [a paper](http://arxiv.org/abs/1004.5168) by Cormack, Smucker and Clarke.

#### Training a Spam Classification Models
To build a spam classification model, you will start with a training data set.   Each instance in the training set represents a single document, and is labeled to indicate whether that document should be considered to be spam or ham.
An instance looks like this:
```
clueweb09-en0094-20-13546 spam 387908 697162 426572 161118 688171 ...
```
The first field, `clueweb09-en0094-20-13546`, is the (unique) document name.   The second field is the label, indicating whether the document should be considered spam (as in this example) or ham.   The remaining fields are integers representing *features* present in the document.   In this case, the features are hashed byte 4-grams, represented as integers.   Each training data set is stored as a text file, with one instance per line.   The training files have been preloaded into the directory `/u/cs451/public_html/spam` for you.   They are:
* `spam.train.group_x.txt`   (25 MB)
* `spam.train.group_y.txt`   (20 MB)
* `spam.train.britney.txt`   (766 MB)

If you are working on your own machine, you can download them using 

    wget https://www.student.cs.uwaterloo.ca/~cs451/spam/spam.train.group_x.txt.bz2
    
from a unix terminal or the Windows bash shell. This will download the compressed file to your current directory. You can extract the files with `bunzip2 spam.train.group_x.txt.bz2`. Follow the instructions at the beginning of [A6-CS451](https://www.student.cs.uwaterloo.ca/~cs451/assignment6-451.html) for more detailed information.

---
### Important

The questions that follow ask you to implement functions whose prototypes are given to you. Do _**not**_ change the prototypes of the functions. Do _**not**_ write code outside of the functions.

You may use specific cells, identified by `# Your tests here`, for test purposes. Code in these cells will *not* be executed when marking your assignment.

---

#### Question 1 ( 5/20 marks)

Your first task is to write a sequential SGD model trainer in Python (no Spark).   For our purposes, a model associates a *weight* with each feature.   The model trainer decides what these weights should be, based on the training instances.  Since you are going to be writing a model trainer based on SGD, the trainer should behave like this:
```
for each training instance T
   predict whether T is spam or ham using the weights of the current model
   update the model weights by comparing T's predicted label with its actual label
```
Of course, the important part is how to update the model.

In [the paper](http://arxiv.org/abs/1004.5168), the model is used to assign a "spamminess" score to a document.   Documents with positive spamminess are predicted to be spam.   Those with negative spamminess are predicted to be ham.  The spamminess of a document $D$ is simply the sum of the weights (from the model) of each of the document's features:
\begin{equation}
spamminess(D) = \sum_{f \in D}{w(f)}
\end{equation}
where $w(f)$ is the weight assocated with feature $f$.

The Python module `spamminess.py` defines a function `spamminess(F,W)` which computes this quantity.   This function takes two arguments, `F` and `W`.  `F` is a list of features (integers) associated to the document whose spamminess you want to compute, and `W` is a dictionary representing the current model.  `W` maps features ($f$) to their weights ($w(f)$) under the model.

In the cell below, you will find partial pseudo-code that shows how to implement the SGD model trainer defined by Cormack, Smucker, and Clarke.   It reads the training instances one at a time from one of the training files, and uses them to adjust the model weights.   Your job is to turn this pseudo-code into actual runnable Python code that can
be used to learn a model from any one of the training files. Implement the function `sequential_SGD()` that takes as input a model (`w`), the training dataset and a value for the update parameter `delta`, and returns the trained model.

In [4]:
from spamminess import spamminess
from math import exp

def sequential_SGD(model, training_dataset='/u/cs451/public_html/spam/spam.train.group_x.txt', delta = 0.002):
    #### Your solution to Question 1 here
    # open one of the training files - defaults to group_x
    with open(training_dataset) as f:
        pass
        #w = dict()
        for line in f:
            
            new = list(line.split(" "))
            
    #      each line represents a document
    #      read and parse the line
            
    #      Let:
    #        t represent the spam/ham tag for this document
    #        F represent the list of features for this document
            t = new[1]
            F = new[2:]
    #      find the spamminess of the current document using the current model:
            score = spamminess(F,model)
            prob =  1.0/(1+exp(-score))
    #      then, update the model:
    #      prob = 1.0/(1+exp(-score))
            for f in F:
                try:
                    model[f]
                except:
                    model[f] =0
                if t == 'spam':
                    model[f] += (1.0-prob)*delta
                elif t == 'ham':
                    model[f] -= 1*prob*delta
                else:
                    model[f] = -999999999999
 
    return model


In [18]:
# Your tests here


w1 = sequential_SGD({}) # Providing an empty model

print("test from the values given in the next question ")
print()
print(w1['271525'] -0.0013133437007968654)
print(w1['802123'] -0.0009858585991850937)
print(w1['438450'] -4.267897922108138e-05)
print(w1['92853'] -0.0004300009932503611)


test from the values given in the next question 

2.60383236130304e-07
3.636757246211486e-06
9.893193150962152e-08
3.313608601634814e-07


#### Question 2 (5/20 marks)

Next, you should try implementing a Spark version of the SGD model trainer.   Before you start, create a new folder
called `models` inside your cs431 folder.   Your Spark implementation should read a training file, train the model, and then output the model to the models folder.  The model output file that you generate should list the weight associated with each feature, with one feature per line, like this:
```
(802123, 0.0009858585991850937)
(438450, 4.267897922108138e-05)
(271525, 0.0013133437007968654)
(92853, 0.0004300009932503611)
```

Use Spark's `saveAsTextFile` action to output your model.   For example, if you are training a model for the group_x training set, use `saveAsTextFile("models/group_x_model")`.   This will actually cause Spark to create a folder called `group_x_model`.   In the folder, there will be files with names like `part-00000` that contain the actual output data.  When you use `saveAsTextFile`, Spark will generate one `part-xxxxx` file for each partition of the RDD that you are writing out.   In this case, you should have only a single partition (for the reason described below), so there should be only one `part-xxxxx` file.

Training the SGD model is an inherently sequential task, since the training instances update the model one at a time, and each instance's spamminess is computed using the model produced by that instance's predecessors.   This means that the only part of the training that you can parallelize using Spark is the parsing of the input file.   Once the input is parsed, your Spark implementation will have to force all of the instances into a single partition, and then apply the training function to the entire partition.   To see whether you are getting sensible results, you can compare the model you learn with Spark to the one that you learned with your sequential Python program from Question 1.

Remember that training should occur entirely in Spark.  The training instances should never come into your driver program.

Implement the function `spark_SGD()` below that takes as input the path to the training dataset, an output path `output_model` and a value for the update parameter `delta`, and writes the trained model to `output_model` using Spark's `saveAsTextFile`. You can use it to generate models from all three of the training files, leaving the results in your models folder. For this assignment, you will be using Spark's original RDD interface, rather than the DataFrame interface.


In [59]:
from spamminess import spamminess
from math import exp
import shutil, os






def spark_SGD(training_dataset='/u/cs451/public_html/spam/spam.train.group_x.txt', output_model='models/group_x_model', delta = 0.002):
    if os.path.isdir(output_model):
        shutil.rmtree(output_model) # Remove the previous model to create a new one
    training_data = sc.textFile(training_dataset)
    def iterators(c, delta = 0.002):
        #splits = list(c.split(","))
        model= {}
        for sen in c:

            new = list(sen.split(" "))
            t = new[1]
            F = new[2:]



            score = spamminess(F,model)
            prob =  1.0/(1+exp(-score))
            for f in F:
                if f not in model:
                    model[f]=0
                if t == 'spam':
                    model[f] += (1.0-prob)*delta
                elif t == 'ham':
                    model[f] -= 1*prob*delta
        mod_list = [(k, v) for k, v in model.items()] 
        return mod_list
    #### Your Solution to Question 2 here
    tr = training_data#sc.textFile('/u/cs451/public_html/spam/spam.train.group_x.txt')
    tr = tr.coalesce(1)
    tr = tr.glom()

    model = tr.flatMap(iterators)
    model = model.map(lambda x: (int(x[0]), x[1]))
    model.saveAsTextFile(output_model)
    #model = model.collect()
    #dict_model = model[0]

    pass
    #return model


In [60]:
# Your tests here
input_model = 'models/group_x_model'
import ast
spark_SGD()
w2 = sc.textFile(input_model).map(lambda x: ast.literal_eval(x.replace('"', '')))
#spark_SGD(training_dataset='/u/cs451/public_html/spam/spam.train.group_y.txt', output_model='models/group_y_model')
#spark_SGD(training_dataset='/u/cs451/public_html/spam/spam.train.britney.txt', output_model='models/britney_model')
w2.take(5)


[(387908, 0.00695474119501872),
 (697162, 0.00695474119501872),
 (426572, 0.00695474119501872),
 (161118, 0.00695474119501872),
 (688171, 0.00695474119501872)]

In [19]:
test = list()
w2 = dict(w2.collect())
a = [key for key in w1.keys() if key in w2.keys()]
a


#comparison of model from question 1 and 2

for i in a:
    k =abs(w1[i]-w2[i])
    test.append(k)

test.sort(reverse = True)
test[0:20]

[0.005449553093766073,
 0.004398510520054098,
 0.0042050794535140905,
 0.0038256543614899707,
 0.003800942012850951,
 0.001972768543173812,
 0.001962007974635165,
 0.0019132971441951944,
 0.0019017255426481295,
 0.0018947580769234746,
 0.0018891466834110738,
 0.0018211573962808522,
 0.0017997452665845643,
 0.0017967137554830736,
 0.0017958960468074537,
 0.001606585363326525,
 0.0015948715875650625,
 0.001528080631803002,
 0.0014975145395184257,
 0.0013162440138839234]

Note that the biggest difference in the weights are in the 3rd decimal place, hence, the two models are quite similar to each other

#### Question 3 (5/20 marks)

When you train a model using SGD, the model you get depends on the order in which you handle the training instances.  To see this in action, try using the Spark SGD trainer you implemented for Question 2 to train a model from the group_x training set, but with the instances processed in a different order.  

To do this, re-implement your trainer from Question 2 so that it will randomly reorder the training instances before using them to update the model. One way to shuffle the training instances is to assign a random sort key to each training instance as you read it from the input file, and then sort the instances using the random sort key.

Be sure that Spark is doing the work of shuffling the training instances.   Do not load the training instances into your driver program and sort them there.

Implement the function `spark_shuffled_SGD` below that takes as input the path to the training dataset, an output path `output_model` and a value for the update parameter `delta`, shuffles the training instances using the method described above and writes the trained model to `output_model` using Spark's `saveAsTextFile`.

Once you have implemented the shuffled trainer, train a model using shuffled group_x training instances, and compare the resulting model with group_x model you learned without shuffling.  It is up to you how to do this comparision.  At a minimum, compare features with the highest weights in each model to see if they are similar. You can also use the classifier in next question to classify documents using the two models, and compare results.


In [61]:
from spamminess import spamminess
from math import exp
import shutil, os, random
import random
import os
import numpy as np




def spark_shuffled_SGD(training_dataset='/u/cs451/public_html/spam/spam.train.group_x.txt', output_model='models/group_x_model', delta = 0.002):
    if os.path.isdir(output_model):
        shutil.rmtree(output_model) # Remove the previous model to create a new one
    def rand(c, n = 756):
    
        l = list(range(756))
        random.shuffle(l)
        return (l[c[1]], c[0])
    tr = sc.textFile(training_dataset)
    #### Your Solution to Question 3 here
    tr = tr.zipWithIndex().map(rand).sortByKey().map(lambda x : x[1])
    tr = tr.coalesce(1)
    tr = tr.glom()

    model = tr.flatMap(iterators).map(lambda x: (int(x[0]), x[1]))
    model.saveAsTextFile(output_model)
    
    pass
    




In [62]:
spark_shuffled_SGD(output_model='models/group_x_model_shuffled')

In [57]:
# Your tests here

results = 'results/test_qrels'

#spark_shuffled_SGD(output_model='models/group_x_model_shuffled')


spark_classify(input_model='models/group_x_model_shuffled')
a = sc.textFile(results).map(lambda x: ast.literal_eval(x.replace('"', '')))
a = a.collect()
a.sort(key = lambda x: x[0])  

spark_classify()
b = sc.textFile(results).map(lambda x: ast.literal_eval(x.replace('"', '')))
b = b.collect()
b.sort(key = lambda x: x[0])  


tot = 0
same = 0
key_err = 0

for i in range(len(a)):
    tot += 1
    if a[i][0] == b[i][0]:
        if a[i][3] == b[i][3]:
            same +=1
    else:
        key_err +=1

print(round(100*same/tot, 2) , "% accuracy")
print(round(100*key_err/tot, 2) , "% error")

c = sc.textFile("models/group_x_model").map(lambda x: ast.literal_eval(x.replace('"', ''))).collect()
d = sc.textFile("models/group_x_model_shuffled").map(lambda x: ast.literal_eval(x.replace('"', ''))).collect()
spark_shuffled_SGD(output_model='models/group_x_model_shuffled')
c.sort(key = lambda x: x[1], reverse = True)
d.sort(key = lambda x: x[1], reverse = True)


for i in range(10):
    print(c[i])
    print(d[i])
    print()

77.53 % accuracy
0.0 % error
(288281, 0.022996007768337472)
(658098, 0.02344231872027235)

(316070, 0.02178760768974702)
(293364, 0.02344231872027235)

(305568, 0.02166991395934579)
(675415, 0.021955556206256888)

(737304, 0.02155627086676939)
(896949, 0.021740971030947)

(102264, 0.021381723804112546)
(756969, 0.02159215685960891)

(202131, 0.020560097038631847)
(275149, 0.021074672902022206)

(674600, 0.02024316823828206)
(288281, 0.020160373574525943)

(419514, 0.0200693349931559)
(520474, 0.020106561974121452)

(837735, 0.019975296559266612)
(202131, 0.01973874896365156)

(879817, 0.01985728645880699)
(718475, 0.019693183356417815)



77.53% of the documents are classified same by both the original and the shuffled SGD. Further it can be seen that the highest weights in both the models are different for the top 10 weights.

#### Question 4 (5/20  marks)

Last but not least, you should write a Spark program that can be used to classify documents as spam or ham, using the classification models you produced.

The test data, i.e., the document instances that you should classifiy, are located in the file
* `/u/cs451/public_html/spam/spam.test.qrels.txt`

Each line in this file represents a document that needs to be classified as spam or ham.  The format of this file is identical to the format of the files that hold the training instances.

Implement the function `spark_classify` below that will load a model (from a specified folder under `models`), classify all of the instances in a given test data file (`spam.test.qrels.txt` by default) using that model, and then output the results in the folder `results_path` using Spark's `saveAsTextFile`.   The contents of the output file should look like this:
```
(clueweb09-en0000-00-00142,spam,2.601624279252943,spam)
(clueweb09-en0000-00-01005,ham,2.5654162439491004,spam)
(clueweb09-en0000-00-01382,ham,2.5893946346394188,spam)
```
Each line of the output represents one test instance.   The first two fields are the document ID and the test label.  These are just copied from the test data.   The third field is the spamminess score of the document, produced by the spamminess function using the model you are classifying with.   The fourth field is the spam/ham prediction made by the model.

Of course, your spam/ham classifier must **not** use the test label from the input when making its prediction.  The test labels are the "ground truth" against which your predictions are being compared.   Using them to make predictions would defeat the whole purpose of model-based classification.

Make sure that classification of the test instances is done by Spark, not by your driver program.  Do ***not*** load the test instances or classification results into your driver program. You are however allowed to load the model weights into your driver program to distribute them as side data. 
Unlike model training, classification is easily parallelizable, since each document is classified independently. 

In [63]:
from spamminess import spamminess
import shutil, os
import ast


def spark_classify(input_model='models/group_x_model', test_dataset='/u/cs451/public_html/spam/spam.test.qrels.txt', results_path='results/test_qrels'):
    if os.path.isdir(results_path):
        shutil.rmtree(results_path) # Remove the previous results
    def classify(c, w ):
        new = list(c.split(" "))
        F = new[2:]
        val = 0
        for k in F:
            f = int(k)
            try:
                w[f]
            except:
                w[f]=0
            val += w[f]
        if val >= 0:
            lab = "spam"
        else:
            lab = "ham"
        return (new[0], new[1], val, lab)
    test_data = sc.textFile(test_dataset)
    
    model = sc.textFile(input_model)
    w = sc.textFile(input_model).map(lambda x: ast.literal_eval(x.replace('"', '')))
    w = dict(w.collect())
    final = test_data.map(lambda x: classify(x,w))
    final.saveAsTextFile(results_path)
    pass




We have installed a program in the CS451 account that can be used to evaluate your classification results.  Given your ouput file, in the proper format, it will compute the area under the receiver operating curve (ROC).   This is a common way to characterize classifier error.    The lower this score, the better.   The evaluation program should produce one line of output, like this
```
1-ROCA%: 17.25
```

Use your classifier to classify the test instances using each of the three classification models that you produced, which should result in three different output files.   Then, in the cell below,
use the evaluation program to evaluate your results.

If you are working on your own machine, you can follow the instructions at the beginning of [CS451 A6](https://www.student.cs.uwaterloo.ca/~cs451/assignment6-451.html) to download and install the program. 

In [68]:
# Your tests here
#  Run the evaluation program like this, after first replacing "output-file"
#  with the name of the folder that holds your classifier's output
#  Note the "!" character, which is important.   This is the escape character
#  that tells the notebook to run an external program.

###############################################################
#commented below is the way to get accuracy of the predictions.
###############################################################

#error = final.map(lambda x: 1 if x[1] == x[3] else 0)
#a = error.collect()
#b = sum(a)
#c = len(a)
#print(b)
#print(c)
#print(b/c)

spark_classify()

#results_path='results/test_qrels'
!/u/cs451/bin/spam_eval.sh 'results/test_qrels' #output-file



1-ROCA%: 17.26


---
Don't forget to save your workbook!   When you are finished and you are ready to submit your assignment, download your notebook file (.ipynb) from the hub to your machine, and then follow the submission instructions in the assignment.